src/dune_rules/pkg_rules.ml

```diff
@@ -1937,6 +1938,26 @@ let build_rule context_name ~source_deps (pkg : Pkg.t) =
        ~directory_targets:[ pkg.write_paths.target_dir ]
 ;;
 
+let install_packages_from_universe ctx_name =
+  let+ packages =
+    (* Fetching the package target implies that we wuill also fetch the extra sources. *)
+    Package_universe.lock_dir (Project_dependencies ctx_name)
+    >>| (fun lock_dir -> lock_dir.packages)
+    >>| Dune_lang.Package_name.Map.keys
+  in
+  let deps =
+    List.map
+      ~f:(fun pkg ->
+        Paths.make ~relative:Path.Build.relative (Project_dependencies ctx_name) pkg
+        |> Paths.target_dir
+        |> Path.build)
+      packages
+    |> Action_builder.paths
+  in
+  Action_builder.bind deps ~f:(fun () ->
+    Action_builder.return (Action.Full.make (Action.Progn [])))
```

Comment:
I think this could be a bit clearer using the operators of `Action_builder.O`.

Reply:
You can use parenthesis for scoping operators like so (or similar, I'm just typing into a text field):

```ocaml
(let open Action_builder.O in
deps >>> Action.Full.make (Action.Progn []))
```

This is one way where the scoping rules of OCaml work really well (looking at you, Python).